### PR TITLE
Add backup verification apis

### PR DESCRIPTION
**Pull request by ishtiaqhimel**

### CODE REVIEW

**Review by anisurrahman75 — CHANGES_REQUESTED** [review]

File: apis/core/v1alpha1/backupconfiguration_types.go [hunk 1]
```diff
@@ -291,31 +295,268 @@ type RepositoryInfo struct {
...
-	// VerifyEvery specifies the frequency of backup verification
-	// +kubebuilder:validation:Minimum=1
-	VerifyEvery int32 `json:"verifyEvery,omitempty"`
+	// VolumeMounts specifies the mount for the volumes specified in `Volumes` section
+	VolumeMounts []core.VolumeMount `json:"volumeMounts,omitempty"`
 
 	// OnFailure specifies what to do if the verification fail.
 	// +optional
-	OnFailure FailurePolicy `json:"onFailure,omitempty"`
+	// OnFailure FailurePolicy `json:"onFailure,omitempty"`
```
Comment (anisurrahman75) [inline apis/core/v1alpha1/backupconfiguration_types.go:324]: What about the `FailurePolicy` field?

File: apis/core/v1alpha1/backupconfiguration_types.go [hunk 2]
```diff
@@ -291,31 +295,268 @@ type RepositoryInfo struct {
...
 	// +optional
-	Params *runtime.RawExtension `json:"params,omitempty"`
+	// KeepAlive *metav1.Time `json:"keepAlive,omitempty"`
+
+	// Function specifies the name of a Function CR that defines a container definition
+	// which will execute the verification logic for a particular application.
+	Function string `json:"function,omitempty"`
+
+	// Volumes indicates the list of volumes that should be mounted on the verification job.
+	Volumes []ofst.Volume `json:"volumes,omitempty"`
```
[the remark this answers is not in the document]
Comment (ishtiaqhimel) [reply apis/core/v1alpha1/backupconfiguration_types.go:317]: We have difference in `EphemeralVolumeSource`. It contains `PartialObjectMeta` for `PersistentVolumeClaimTemplate`.

File: apis/core/v1alpha1/backupconfiguration_webhook.go [hunk 3]
```diff
@@ -429,6 +441,52 @@ func targetMatched(t1, t2 *kmapi.TypedObjectReference) bool {
...
+
+		if vs.RestoreOption.AddonInfo == nil {
+			return fmt.Errorf("addonInfo in restoreOption for verification strategy %q cannot be empty", vs.Name)
+		}
+
+		if vs.VerifySchedule == "" {
+			return fmt.Errorf("verify schedule for verification strategy %q cannot be empty", vs.Name)
+		}
+
+		if vs.File != nil || vs.Query != nil || vs.Script != nil {
```
[the remark this answers is not in the document]
Comment (ishtiaqhimel) [reply apis/core/v1alpha1/backupconfiguration_webhook.go:466]: We need a `Function` for these types. For `RestoreOnly` we don't need any function.

File: apis/core/v1alpha1/backupconfiguration_types.go (hunk 1 above)
Comment (ishtiaqhimel) [reply apis/core/v1alpha1/backupconfiguration_types.go:324]: We still don't have any plan regarding `FailurePolicy`. You can find that we also removed this field from backup and restore.

**Review by anisurrahman75 — APPROVED** [review]

File: apis/constant.go [hunk 4]
```diff
@@ -51,16 +51,18 @@ const (
...
-	KubeStashHookComponent        = "kubestash-hook"
-	KubeStashPopulatorComponent   = "kubestash-populator"
+	KubeStashBackupComponent             = "kubestash-backup"
+	KubeStashRestoreComponent            = "kubestash-restore"
+	KubeStashInitializerComponent        = "kubestash-initializer"
+	KubeStashUploaderComponent           = "kubestash-uploader"
+	KubeStashCleanerComponent            = "kubestash-cleaner"
+	KubeStashHookComponent               = "kubestash-hook"
+	KubeStashPopulatorComponent          = "kubestash-populator"
+	KubeStashBackupVerificationComponent = "kubestash-backup-verification"
```
Comment (hmsayem) [inline apis/constant.go:65]:
```suggestion
	KubeStashBackupVerifierComponent = "kubestash-backup-verifier"
```

File: apis/constant.go [hunk 5]
```diff
@@ -100,13 +102,14 @@ const (
...
-	KubeStashPopulatorJobClusterRole       = "kubestash-populator-job"
-	KubeStashRetentionPolicyJobClusterRole = "kubestash-retention-policy-job"
+	KubeStashBackupJobClusterRole             = "kubestash-backup-job"
+	KubeStashRestoreJobClusterRole            = "kubestash-restore-job"
+	KubeStashCronJobClusterRole               = "kubestash-cron-job"
+	KubeStashBackendJobClusterRole            = "kubestash-backend-job"
+	KubeStashStorageInitializerClusterRole    = "kubestash-storage-initializer-job"
+	KubeStashPopulatorJobClusterRole          = "kubestash-populator-job"
+	KubeStashRetentionPolicyJobClusterRole    = "kubestash-retention-policy-job"
+	KubeStashBackupVerificationJobClusterRole = "kubestash-backup-verification-job"
```
Comment (hmsayem) [inline apis/constant.go:112]:
```suggestion
	KubeStashBackupVerifierJobClusterRole = "kubestash-backup-verifier-job"
```

File: apis/core/v1alpha1/backupverificationsession_types.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	ReasonFailedToEnsureVerificationExecutor      = "FailedToEnsureVerificationExecutor"
+
+	// TypeRestoreSucceeded indicates whether the restore is succeeded or not.
+	TypeRestoreSucceeded             = "RestoreSucceeded"
+	ReasonSuccessfullySucceedRestore = "SuccessfullySucceedRestore"
+	ReasonFailedToSucceedRestore     = "FailedToSucceedRestore"
+
+	// TypeBackupVerified indicates whether backup is verified or not
+	TypeBackupVerified             = "BackupVerified"
+	ReasonSuccessfullyVerifyBackup = "SuccessfullyVerifyBackup"
```
Comment (hmsayem) [inline apis/core/v1alpha1/backupverificationsession_types.go:126]:
```suggestion
	ReasonSuccessfullyVerifiedBackup = "SuccessfullyVerifiedBackup"
```

File: apis/core/v1alpha1/backupverificationsession_types.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+	// TypeVerificationExecutorEnsured indicates whether the backup verification executor is ensured or not.
+	TypeVerificationExecutorEnsured               = "VerificationExecutorEnsured"
+	ReasonSuccessfullyEnsuredVerificationExecutor = "SuccessfullyEnsuredVerificationExecutor"
+	ReasonFailedToEnsureVerificationExecutor      = "FailedToEnsureVerificationExecutor"
+
+	// TypeRestoreSucceeded indicates whether the restore is succeeded or not.
+	TypeRestoreSucceeded             = "RestoreSucceeded"
+	ReasonSuccessfullySucceedRestore = "SuccessfullySucceedRestore"
+	ReasonFailedToSucceedRestore     = "FailedToSucceedRestore"
```
Comment (hmsayem) [inline apis/core/v1alpha1/backupverificationsession_types.go:122]:
```suggestion
	ReasonSuccessfullyRestoredBackup = "SuccessfullyRestoredBackup"
	ReasonFailedToRestoreBackup     = "FailedToRestoreBackup"
```

File: apis/storage/v1alpha1/snapshot_types.go [hunk 8]
```diff
@@ -110,6 +110,10 @@ type SnapshotStatus struct {
 	// +optional
 	VerificationStatus VerificationStatus `json:"verificationStatus,omitempty"`
 
+	// VerificationSession specifies which BackupVerificationSession verified this Snapshot
+	// +optional
+	VerificationSession string `json:"verificationSession,omitempty"`
+
```
Comment (hmsayem) [inline apis/storage/v1alpha1/snapshot_types.go:116]: do we need this field?

**Review by hmsayem — CHANGES_REQUESTED** [review]

File: apis/core/v1alpha1/backupverifier_types.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+// BackupVerifierSpec specifies the information related to the respective restore target, verification schedule, and verification type.
+type BackupVerifierSpec struct {
+	// RestoreOption specifies the restore target, and addonInfo for backup verification
+	// +optional
+	RestoreOption *RestoreOption `json:"restoreOption,omitempty"`
+
+	// Schedule specifies the schedule of backup verification in Cron format, see https://en.wikipedia.org/wiki/Cron.
+	Schedule string `json:"schedule,omitempty"`
```
Comment (hmsayem) [inline apis/core/v1alpha1/backupverifier_types.go:53]: use https://github.com/kubestash/apimachinery/blob/4e9720f4cb0689efd38e46d127e70fcf0184cc5a/apis/core/v1alpha1/backupconfiguration_types.go#L143

**Review by hmsayem — CHANGES_REQUESTED** [review]

**Review by hmsayem — APPROVED** [review]

**Pull request merged**